Bug 1278559 share downloadable files

fluffyemily opened the pull request:
Add ability to share files that cannot be handled by the browser using the share menu

sleroux commented:
Are there any links you were using to test this out? Every .txt page I visit properly renders in the browser :(

fluffyemily commented:
@sleroux I logged into my dropbox account and started clicking on random files. Also, try opening zip files from bugzilla bugs, that also works.

jhugman commented on Client/Frontend/Browser/OpenPdfHelper.swift:
```diff
@@ -26,36 +26,77 @@ enum FileType : String {
+    var fileType: FileType { get }
+    var documentDirectory: NSURL { get }
+
+    static func canOpenType(type: String) -> Bool
```
Would prefer to see `canOpenFileType` and `canOpenMimeType` here.

e.g.
```swift
guard let type = MimeType(rawValue: type) else {
    // …
}
if OpenPassBookHelper.canOpenMimeType(type) {
    // …
}
```

jhugman commented on Client/Frontend/Browser/OpenPdfHelper.swift:
```diff
@@ -26,36 +26,77 @@ enum FileType : String {
-        if OpenPdfInHelper.canOpen(url) {
+
+    static func helperForURL(url: NSURL, mimeType: String? = nil) -> OpenInHelper? {
+        guard let type = mimeType ?? url.pathExtension else {
```
Is this code in use anymore?

sleroux commented on Client/Frontend/Browser/OpenInHelper.swift:
```diff
@@ -24,38 +24,107 @@ struct OpenInViewUX {
-    var openInView: OpenInView { get }
-    static func canOpen(url: NSURL) -> Bool
+    init(url: NSURL)
+    init?(response: NSURLResponse)
```
Nit: Looking below as to how both these `init` calls are being used, I think you can get away with using `init(url; NSURL)` as your designated initializer and making `init(response: NSURLResponse)` a convenience one that calls into `init(url..)`. Similar thing can be done for the `helperForURL`/`helperForResponse` in that you can make one a helper around the other since they both end up only using the URL anyways.

sleroux commented on Client/Frontend/Browser/OpenInHelper.swift:
```diff
@@ -24,38 +24,107 @@ struct OpenInViewUX {
+}
+
+class ShareFileHelper: NSObject, OpenInHelper {
+    lazy var openInView: OpenInView? = nil
```
Nit: Using `lazy` here doesn't do anything since it will lazy load nil into this property and get setup later anyways.

sleroux commented on Client/Frontend/Browser/OpenInHelper.swift:
```diff
@@ -24,38 +24,107 @@ struct OpenInViewUX {
 }
 
 class OpenPassBookHelper: NSObject, OpenInHelper {
+    lazy var openInView: OpenInView? = nil
```
Nit: `lazy`

sleroux commented on Client/Frontend/Browser/OpenInHelper.swift:
```diff
@@ -91,11 +160,34 @@ class OpenPdfInHelper: NSObject, OpenInHelper, UIDocumentInteractionControllerDe
+        return NSURL(string: NSTemporaryDirectory())!.URLByAppendingPathComponent(FileType.PDF.rawValue)
+    }()
+
+    private var filepath: NSURL?
```
Nit: You can probably make this a `let` and non-optional since you have everything you need in the `init` call before calling super to setup this variable.

jhugman commented on Client/Frontend/Browser/OpenPdfHelper.swift:
```diff
@@ -91,11 +132,37 @@ class OpenPdfInHelper: NSObject, OpenInHelper, UIDocumentInteractionControllerDe
+        guard let responseURL = response.URL else { return nil }
+        url = responseURL
+        super.init()
+        setFilePath(response.suggestedFilename ?? url.lastPathComponent ?? "file.pdf")
```
Ok. Gotcha.

I'm bothered by a lot of the code duplication around this constructors. I'm not convinced that both of them are needed.

fluffyemily replied:
Yeah, you're right. We used to only have the URL one, then I created the response one for the general open in case. Once I then moved all of the others over to also use the response initialiser I probably should have just dumped the old url one as it was no longer needed. Probably the same with the file type check too, as the response will always have the MIMEType.

fluffyemily commented:
@sleroux @jhugman I've updated the code to present an alert asking for the users consent before presenting the share sheet. Also addressed some nits by removing the unused init(NSURL) and openInHelperForURL functions.

jhugman commented:
Overall:
 * there's a lot of duplication of code, most of it redundant. I think you've addressed most of that. My suspicion is that the code will get dramatically simpler once you assume `NSURLResponse` (I think you've already started that).
 * the factory is doing lots of if/elsing and constructing. The pseudo code I was thinking of:

For helper protocol:
```
protocol Helper {
  static func canOpen(response: NSURLResponse) -> Bool
  static func open(url: NSURL)
}
```

```
let helpers = [HelperA.self, HelperB.self, HelperC.self]
if let helper = helpers.find { $0.canOpen(response) }.first {
  return helper.open
}
``` 

 * it's quite Javay: the open helper is an object that is only useful as a single function. e.g.

```
typealias URLOpener = (NSURL) -> ()
protocol OpenInFactory {
    func findOpenIn(response: NSURLResponse) -> URLOpener
}
```

fluffyemily commented:
Thing is @jhugman is that not all the OpenInHelpers work the same. In the case of pdfs we need to provide a view that will open when requested and so that helper needs to be kept around until such time as it is called. The more recent open in's fit better with the style that you've stated above, but I don't think it would work in the case of the pdf open in helper.

The whole factory thing could definitely be improved though, and the init doesn't need to be in the protocol. It was there as I was hoping to be able to provide a default implementation, but that doesn't seem to work out so well. I'll see what I can do to address your points while ensuring that we keep the Open Pdf In functionality the same.

flodolo commented on Client/Frontend/Strings.swift:
```diff
@@ -170,7 +170,17 @@ extension Strings {
+extension Strings {
+    public static let OpenInDownloadHelperAlertTitle = NSLocalizedString("Downloads.Alert.Title", value: "Firefox Downloads", comment: "The title of the alert box asking the user if they want to use another app to open a file.")
+    public static let OpenInDownloadHelperAlertMessage = NSLocalizedString("Downloads.Alert.Message", value: "Firefox is unable to download or display this file. Would you like to open it in another app?", comment: "The message of the alert box asking the user if they want to use another app to open a file.")
+    public static let OpenInDownloadHelperAlertConfirm = NSLocalizedString("Downloads.Alert.Confirm", value: "Yes", comment: "The title of the button the user will press to be presented with a list of other apps to open a file in")
```
Label, not title (also the following one).

fluffyemily commented:
@jhugman best I could do for now.

jhugman commented on Client/Frontend/Browser/OpenInHelper.swift:
```diff
@@ -21,41 +21,63 @@ struct OpenInViewUX {
+    
+    static func helperForResponse(response: NSURLResponse) -> OpenInHelper? {
+        return helpers.flatMap { $0.init(response: response) }.first
     }
```
That's really nice.

jhugman commented:
That's so much cleaner. Good job.
